File: samples/php/verify_and_read.php
<?php
// KeyNub dongle check from PHP: enumerate -> open -> verify -> session ->
// read a record -> app-crypto round trip.
//
//   KEYNUB_LICDONGLE_LIBRARY=../../build/keynub_licdongle.dll php verify_and_read.php
//
// Targets real hardware: with no dongle attached it prints guidance and exits 0.
//
// READ FIRST: docs/integration-security.md. This sample prints whether the dongle
// is genuine, which is the one thing a real licence check must not do — a printed
// boolean is a deleted line away from nothing. protectSomething() shows the shape
// that actually protects something.
//
// PHP is the case where the dongle earns its place specifically for *on-premise*
// installs: if you host the application yourself you do not need one, but PHP
// business software is very often installed on a customer's own server, where they
// hold your source and an encoder's licence file is a file, which copies.

declare(strict_types=1);

require __DIR__ . '/../../bindings/php/src/KeyNub.php';

use KeyNub\LicDongle\Context;
use KeyNub\LicDongle\Dongle;
use KeyNub\LicDongle\LicenseDongleException;
use KeyNub\LicDongle\Scope;
use KeyNub\LicDongle\Session;

function report(Dongle $dongle): void
{
    $info = $dongle->getInfo();
    printf(
        "Protocol v%d.%d, firmware v%d.%d.%d, %d of %d bytes free.\n",
        $info->protocolMajor, $info->protocolMinor,
        $info->firmwareMajor, $info->firmwareMinor, $info->firmwarePatch,
        $info->dataFree, $info->dataCapacity
    );

    if ($info->watchdogReboot) {
        // The only trace a firmware hang leaves behind. Worth reporting to support.
        echo "WARNING: this dongle's previous boot ended in a watchdog reset.\n";
    }

    $result = $dongle->verifyGenuine();
    printf(
        "Genuine: %s (serial %s, batch %s, provisioned %s)\n",
        $result->genuine ? 'true' : 'false',
        $result->serial, $result->batch, $result->provisionedDate
    );
}

function readRecords(Session $session): void
{
    $records = $session->listRecords();
    printf("%d record(s) on the dongle:\n", count($records));
    foreach ($records as $record) {
        printf("  %-16s %6d bytes\n", $record->name, $record->size);
    }

    foreach ($records as $record) {
        if ($record->name === 'license') {
            $data = $session->readRecord('license');
            printf("Read %d bytes from the license record.\n", strlen($data));
            break;
        }
    }
}

// The part that actually protects something. At licence-issue time you would call
// appEncrypt once, with a developer dongle, and ship only the blob; the application
// then cannot proceed without a dongle, because it holds no other copy of the data.
// Scope::DEVELOPER lets any dongle from your batch decrypt it, so one file serves
// every customer; Scope::DEVICE locks it to one dongle.
//
// Check on boot or on a schedule rather than per request: a USB round trip is fast
// but not free, and a per-request check buys nothing an hourly one does not.
function protectSomething(Session $session): void
{
    $needed = 'the data this program cannot run without';

    $sealed = $session->appEncrypt(Scope::DEVELOPER, $needed);
    $recovered = $session->appDecrypt($sealed);

    printf(
        "App-crypto round trip: %d bytes -> %d sealed -> %s\n",
        strlen($needed),
        strlen($sealed),
        // hash_equals, not ===: constant-time comparison is the habit worth having
        // anywhere a mismatch is security-relevant.
        hash_equals($needed, $recovered) ? 'recovered intact' : 'MISMATCH'
    );
}

[$major, $minor, $patch] = Context::libraryVersion();
printf("KeyNub SDK %d.%d.%d\n", $major, $minor, $patch);

try {
    $ctx = new Context();

    $devices = $ctx->enumerate();
    printf("Found %d KeyNub dongle(s).\n", count($devices));
    foreach ($devices as $i => $d) {
        printf("  [%d] serial %s (VID %04X PID %04X)\n", $i, $d->serial, $d->vendorId, $d->productId);
    }
    if (count($devices) === 0) {
        echo "No dongle attached; nothing to do.\n";
        exit(0);
    }

    // No argument = first dongle found; pass a serial to pick a specific one.
    $dongle = $ctx->open();
    report($dongle);

    $session = $dongle->openSession();
    readRecords($session);
    protectSomething($session);

    $session->close();
    $dongle->close();
    $ctx->close();
} catch (LicenseDongleException $e) {
    // The exception subclass distinguishes the cases; the detail carries the SDK's
    // diagnostic text, which is what tells "no dongle" from "certificate rejected".
    fprintf(STDERR, "KeyNub error: %s\n", $e->getMessage());
    exit(1);
}

File: samples/lua/verify_and_read.lua
-- KeyNub dongle check from Lua: enumerate -> open -> verify -> session ->
-- read a record -> app-crypto round trip.
--
--   KEYNUB_LICDONGLE_LIBRARY=../../build/keynub_licdongle.dll \
--     luajit -e "package.path='../../bindings/lua/?.lua;'..package.path" verify_and_read.lua
--
-- Requires LuaJIT: its ffi module is what makes the binding one file with nothing
-- to compile.
--
-- Targets real hardware: with no dongle attached it prints guidance and exits 0.
--
-- READ FIRST: docs/integration-security.md. This sample prints whether the dongle
-- is genuine, which is the one thing a real licence check must not do -- a printed
-- boolean is a deleted line away from nothing. protectSomething shows the shape
-- that actually protects something, and Lua is where it matters most: a plugin
-- usually ships as readable source alongside the application, so
-- `if not licensed then os.exit() end` is one line of a file the customer can edit.

local ok, keynub = pcall(require, 'keynub_licdongle')
if not ok then
  -- Running from a checkout, where the rock is not installed.
  local dir = debug.getinfo(1, 'S').source:match('^@(.*)[/\\][^/\\]*$') or '.'
  package.path = dir .. '/../../bindings/lua/?.lua;' .. package.path
  keynub = require('keynub_licdongle')
end

local function report(dongle)
    local info = dongle:getInfo()
    print(string.format('Protocol v%d.%d, firmware v%d.%d.%d, %d of %d bytes free.',
        info.protocolMajor, info.protocolMinor,
        info.firmwareMajor, info.firmwareMinor, info.firmwarePatch,
        info.dataFree, info.dataCapacity))

    if info.watchdogReboot then
        -- The only trace a firmware hang leaves behind. Worth reporting to support.
        print("WARNING: this dongle's previous boot ended in a watchdog reset.")
    end

    local result = dongle:verifyGenuine()
    print(string.format('Genuine: %s (serial %s)',
        tostring(result.genuine), result.serial))
end

local function readRecords(session)
    local records = session:listRecords()
    print(string.format('%d record(s) on the dongle:', #records))
    for _, record in ipairs(records) do
        print(string.format('  %-16s %6d bytes', record.name, record.size))
    end

    for _, record in ipairs(records) do
        if record.name == 'license' then
            local data = session:readRecord('license')
            -- #data is correct for binary here: byte data is a Lua string throughout.
            print(string.format('Read %d bytes from the license record.', #data))
            break
        end
    end
end

-- The part that actually protects something. At licence-issue time you would call
-- appEncrypt once, with a developer dongle, and ship only the blob; the add-on then
-- cannot proceed without a dongle, because it holds no other copy of the data.
-- Scope.DEVELOPER lets any dongle you have issued decrypt it, so one file serves
-- every customer; Scope.DEVICE locks it to one dongle.
local function protectSomething(session)
    local needed = 'the data this program cannot run without'

    local sealed = session:appEncrypt(keynub.Scope.DEVELOPER, needed)
    local recovered = session:appDecrypt(sealed)

    print(string.format('App-crypto round trip: %d bytes -> %d sealed -> %s',
        #needed, #sealed, recovered == needed and 'recovered intact' or 'MISMATCH'))
end

local function run()
    local ctx = keynub.Context()

    local devices = ctx:enumerate()
    print(string.format('Found %d KeyNub dongle(s).', #devices))
    for i, d in ipairs(devices) do
        print(string.format('  [%d] serial %s (VID %04X PID %04X)',
            i - 1, d.serial, d.vendorId, d.productId))
    end
    if #devices == 0 then
        print('No dongle attached; nothing to do.')
        ctx:close()
        return
    end

    -- No argument = first dongle found; pass a serial to pick a specific one.
    local dongle = ctx:open()
    report(dongle)

    local session = dongle:openSession()
    readRecords(session)
    protectSomething(session)

    session:close()
    dongle:close()
    ctx:close()
end

local major, minor, patch = keynub.libraryVersion()
print(string.format('KeyNub SDK %d.%d.%d', major, minor, patch))

-- Errors are raised as tables with status/operation/message/detail and a
-- __tostring, so this stays readable whichever way it arrives.
local ok, err = pcall(run)
if not ok then
    io.stderr:write('KeyNub error: ' .. tostring(err) .. '\n')
    os.exit(1)
end
